Wait for memory units on obligatory batch reads in fetch handler

ballard26 opened the pull request:
Prior to this PR obligatory batch reads would be carried out regardless of available memory. This lead to OOMs in cases of lots of pending requests.

This PR changes obligatory read behavior to wait for memory units if none are available. If after `max.wait.ms` not enough units are available still then the obligatory batch read will not be made. In cases of memory being starved on every shard this can lead to the client getting a response with no data.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

ballard26 committed:
kafka/server: wait for memory units on obligatory batch reads

Prior to this commit obligatory batch reads would be carried out
regardless of available memory. This lead to OOMs in cases of lots of
pending requests.

This commit changes obligatory read behavior to wait
for memory units if none are available. If after `max.wait.ms` not
enough units are available still then the obligatory batch read will not
be made. In cases of memory being starved on every shard this can lead
to the client getting a response with no data.

ballard26 commented on src/v/kafka/server/fetch_memory_units.cc:
```diff
@@ -117,12 +130,43 @@ void fetch_memory_units_manager::release_units_to_semaphore(units&& u) {
+        // Try to get the fetch units first. They should normally be the lesser
+        // of the two and holding them won't block other request types on
+        // this shard.
+        auto fetch_units = co_await ss::get_units(_fetch_units, target, as);
+        auto kafka_units = co_await ss::get_units(_kafka_units, target, as);
```
`fetch_units` is RAII and the units will be returned to the semaphore whenever the object is destroyed.

ballard26 commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -343,15 +341,18 @@ ss::future<read_result> read_from_ntp(
   fetch_memory_units_manager& units_mgr) {
-    return do_read_from_ntp(
+    ss::abort_source as{};
+    as.request_abort();
```
Using a non-aborted source here can caused indefinite hanging in cases where there isn't enough memory units for an obligatory read. Furthermore no existing fetch test using this function relies on obligatory batch reads.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74124-->
<!--048bac75-ad1b-4603-bb8c-3af491c4cbb0-->
<!--e0372341-600e-490f-aabe-719b6aa014bc-->
<!--2da57b99-10fc-43d4-94d4-a86d30208cbf-->
<!--b819bfc1-6bc6-4090-89d5-befe60fdca46-->
<!--5c17a905-948e-4d9b-bb09-f500b89872bb-->
<!--077f7e21-c33a-491b-a514-a57ab0840834-->
<!--598a6000-05f3-41a8-8ec3-dddd6536a2f5-->
<!--3d73c519-4432-4306-81f4-117bb943bf5b-->
<!--f542969e-c07e-483b-93e7-045094a19e3d-->
<!--8a2efdac-d1de-4c24-a9e7-98f761df65d8-->
<!--d7e21bab-2dd6-43aa-8f65-b541c675912c-->
<!--9d8fb029-b671-4ad1-8091-10a0b085c639-->
<!--title-->
## Retry command for [Build#74124](https://buildkite.com/redpanda/redpanda/builds/74124)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_client_response_throttle_mechanism_applies_to_next_request
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_client_quota_metrics
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_throttling_ms_enforcement_is_per_connection
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_client_response_and_produce_throttle_mechanism
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_client_response_throttle_mechanism
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_client_group_consume_rate_throttle_mechanism
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1eb9df12-5e4e-44a6-989d-b9b16310e018-->
<!--TEST RESULTS-->
<!--142494d5-5bc8-4e57-96ff-df66814237ab-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74124">build#74124</a></summary>


| test_class           | test_method                                                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                                                       |
|:---------------------|:----------------------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterRateQuotaTest | test_client_group_consume_rate_throttle_mechanism               | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e127-73f3-40a1-9379-8f5d0f9cbec3 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism               |
| ClusterRateQuotaTest | test_client_group_consume_rate_throttle_mechanism               | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e128-03c8-4be9-9890-27ef65f89dec | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism               |
| ClusterRateQuotaTest | test_client_quota_metrics                                       | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e127-73f5-4694-8f1a-86852823b224 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_quota_metrics                                       |
| ClusterRateQuotaTest | test_client_quota_metrics                                       | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e128-03cb-4f25-81f5-1c3446cde42d | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_quota_metrics                                       |
| ClusterRateQuotaTest | test_client_response_and_produce_throttle_mechanism             | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e127-73f6-42b6-8627-97502d066ef3 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest | test_client_response_and_produce_throttle_mechanism             | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e128-03cd-47a6-b4db-488ee381a7c8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest | test_client_response_throttle_mechanism                         | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e127-73f6-4df6-a847-f70d614e99a8 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest | test_client_response_throttle_mechanism                         | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e128-03d0-45df-a776-d8344d36f1a0 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest | test_client_response_throttle_mechanism_applies_to_next_request | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e127-73f7-4dd4-b342-818a62863cf3 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| ClusterRateQuotaTest | test_client_response_throttle_mechanism_applies_to_next_request | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e128-03d2-4a60-89ce-736488610049 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| ClusterRateQuotaTest | test_throttling_ms_enforcement_is_per_connection                | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e127-73f0-4f20-a6c9-098d1308a5a6 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_throttling_ms_enforcement_is_per_connection                |
| ClusterRateQuotaTest | test_throttling_ms_enforcement_is_per_connection                | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74124#0199e128-03c0-443a-8218-676ea8acd8bb | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_throttling_ms_enforcement_is_per_connection                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74912">build#74912</a></summary>


| test_class                      | test_method                          | test_arguments                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:--------------------------------|:-------------------------------------|:---------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest       | test_write                           | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/74912#019a1699-f180-41e9-8512-01d76873998a | FLAKY         | 19/21    | upstream reliability is '98.0603448275862'. current run reliability is '90.47619047619048'. drift is 7.58415 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                       |
| DataMigrationsApiTest           | test_creating_and_listing_migrations | null                                               | integration | https://buildkite.com/redpanda/redpanda/builds/74912#019a1699-f189-41f3-bd63-c608e23ddb66 | FLAKY         | 17/21    | upstream reliability is '99.8533724340176'. current run reliability is '80.95238095238095'. drift is 18.90099 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| NodesDecommissioningTest        | test_recommissioning_node            | {"cloud_topic": true}                              | integration | https://buildkite.com/redpanda/redpanda/builds/74912#019a16a4-8bdc-4324-ae17-49db7dba683e | FLAKY         | 19/21    | upstream reliability is '98.86039886039886'. current run reliability is '90.47619047619048'. drift is 8.38421 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node         |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test             | {"cloud_storage_type": 1, "mixed_versions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74912#019a16a4-8bd4-4f9a-8056-c4a598060c25 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
</details>
<!--/body-->

ballard26 commented:
So the `ClusterRateQuotaTest` failures are a bit interesting. The `message.max.bytes` for the topic is being set to 1GiB. This results in every read to a partition in that topic trying to reserve 1GiB in units from the fetch and kafka semaphore. Prior to this PR consumers would still make progress as the semaphore units were allowed to go negative to satisfy an obligatory batch read. Now, however, this is not the case and since it's impossible to allocate 1GiB in memory units for the CI runs of this test the consumers make no progress. Hence the test failures.

Here's the total units we actually have per shard in CI: 
```
INFO  2025-10-14 06:17:33,856 [shard 0:main] main - Per shard memory group allocations: total memory: 2.000GiB, total memory minus pre-share reservations: 1.675GiB, chunk cache: 302.682MiB, kafka: 605.365MiB, rpc: 403.576MiB, recovery: 201.788MiB, tiered storage: 201.788MiB, data transforms: 0.000bytes, compaction: 128.000MiB, datalake: 0.000bytes, partitions: 204.800MiB
``` 

The test can be changed to pass with this PR. The failures though raise a question about how tolerant Redpanda should be of potentially misconfigured clients. I.e, if there aren't enough units for `message.max.bytes` on the obligatory read maybe we could allocate as much units as possible and see if it's possible to read the next batch with those. Otherwise we could fallback to trying to allocate `message.max.bytes`.  @BenPope recommended similar before IIRC.

pgellert commented on src/v/kafka/server/tests/fetch_test.cc:
```diff
@@ -1210,3 +1210,102 @@ FIXTURE_TEST(fetch_response_bytes_eq_units, redpanda_thread_fixture) {
+
+    BOOST_REQUIRE(fetch_sem.available_units() == 0);
+
+    // The first response is in a state where it hasn't been sent to the client
+    // yet. Hence it is still holding all available fetch semaphore units.
+    // Therefore this response should contain no data.
+    auto octx_2 = kafka::op_context(
+      make_rctx(), ss::default_smp_service_group());
+    kafka::testing::do_fetch(octx_2).get();
+    BOOST_REQUIRE(octx_2.response_size == 0);
```
Is it safe to return no data to the clients without an error code in this case? I assume it is because the client will just treat it as if there was genuinely no data and retry.

ballard26 replied:
Yeah that's my assumption as well. The offsets we'd return should still show there is more to read from that ntp so I'd expect the client to just retry. 

@twmb any input here would be much appreciated. The question is basically what Redpanda should return to clients in cases where we can't read from an ntp due to low memory. The currently implemented solution is to return no batches(and no error) along with the most up-to-date offsets for an ntp that was skipped due to memory constraints.

twmb replied:
I think this will mirror the [KAFKA-5443](https://issues.apache.org/jira/browse/KAFKA-5443) case and should be fine, OR it will mirror the case where the client is requesting at the end offset and no new data is available (so, no batches are returned). But it's _definitely_ worth testing.

pgellert commented on src/v/kafka/server/handlers/fetch.cc:
```diff
@@ -221,24 +221,22 @@ static ss::future<read_result> do_read_from_ntp(
+          obligatory_batch_read,
+          as);
+        if (!memory_units.has_units()) {
+            ntp_config.cfg.skip_read = true;
```
To check my understanding, if we can't get enough units for the first partition, then we skip it but continue to try to read from other partitions, right?

ballard26 replied:
Yes, however, since the abort source would've been aborted after failing to get units from the first partition. This would mean that we'd be unlikely to get units for the rest(though a non-waiting attempt is made). Furthermore the `octx.deadline` would've been hit at this point as well(assuming `fetch.max.wait.ms > 0`). And I'm unsure if the storage layer will try to read data from the partition if the deadline on the reader config was already hit.

pgellert replied:
> This would mean that we'd be unlikely to get units for the rest(though a non-waiting attempt is made).

This is true on the same shard, but not necessarily on other shards, right? And this is true if the fetch is for a single topic, but not necessarily if the fetch is for multiple topics with different `max.message.bytes` (although perhaps that is a rare/unexpected scenario), right?

ballard26 replied:
> This is true on the same shard, but not necessarily on other shards, right?

Yeah, each shard will run `fetch_ntps` independently of one another. This also means there will be one obligatory read per shard. Not optimal, but it is the best fit with RP/seastar's architecture. 

> And this is true if the fetch is for a single topic, but not necessarily if the fetch is for multiple topics with different max.message.bytes (although perhaps that is a rare/unexpected scenario), right?

On a given shard it's true even if the partitions are from multiple topics. We'll still treat the first non-empty partition in the list as the obligatory read. And therefore will wait for the needed memory units, potentially until aborted. Afterwards non-waiting attempts will be made for the rest of the partitions on that shard.

pgellert commented on src/v/kafka/server/fetch_memory_units.cc:
```diff
@@ -117,12 +130,43 @@ void fetch_memory_units_manager::release_units_to_semaphore(units&& u) {
+ss::future<fetch_memory_units> fetch_memory_units_manager::get_units(
+  const size_t target, ss::abort_source& as) {
+    try {
+        // Try to get the fetch units first. They should normally be the lesser
+        // of the two and holding them won't block other request types on
+        // this shard.
+        auto fetch_units = co_await ss::get_units(_fetch_units, target, as);
+        auto kafka_units = co_await ss::get_units(_kafka_units, target, as);
```
Should we add a counter metric here for the number of times we timed out waiting for fetch/kafka units, so that we can observe it and react by adjusting the max.message.size or similar?

ballard26 replied:
Good idea. Whatever metric is added would likely need a topic label. Would we want just the number of failed waits? Or maybe something like the total time spent waiting? Or both?

Another question is what adjustments can be made to decrease these metrics if identified as an issue. Max message size is one option as you said. However, there are risks to memory safety when decreasing `max.message.size` if a batch greater than the new, smaller, max size was produced to the topic at some point in the past.

pgellert replied:
> Whatever metric is added would likely need a topic label.

Yeah, either that or perhaps a metric + a debug log line to help identify the topic.

> Would we want just the number of failed waits? Or maybe something like the total time spent waiting? Or both?

Yeah total time spent sounds even better. We probably want both then because different consumers could have different max wait ms, so we wouldn't necessarily know from the latency metric how many timeouts we have seen.

> Another question is what adjustments can be made to decrease these metrics if identified as an issue.

Yeah, I guess it depends on the workload. Moving to an instance with more memory (and therefore more fetch semaphores) is another option.

pgellert reviewed (commented):
> if there aren't enough units for message.max.bytes on the obligatory read maybe we could allocate as much units as possible and see if it's possible to read the next batch with those

With this proposal, if we attempt a read with fewer units, would we need to use `strict_max_bytes = true` to avoid an OOM?

ballard26 commented on src/v/kafka/server/fetch_memory_units.cc:
```diff
@@ -64,8 +64,12 @@ fetch_memory_units_manager::units::~units() noexcept {
       "foreign units need to be released via the fetch_memory_units_manager");
```
> With this proposal, if we attempt a read with fewer units, would we need to use strict_max_bytes = true to avoid an OOM?

@pgellert Yeah, this exactly. Basically take as many units as we can without waiting. Try to read from the partition with `strict_max_bytes = true`. Then if that fails to read a batch fallback to waiting on units.

pgellert replied:
Yeah I think that sounds like a good idea. I am wondering if we could be even more optimistic on the first attempt then, rather than taking all the available units, perhaps we could take 2x the `avg_bytes_per_offset` estimate from https://github.com/redpanda-data/redpanda/pull/27909.

I would expect that in general max.message.bytes will be >> the typical batch sizes seen on the topic, just because of the nature of this config. I'd expect that customers will probably typically set it at least an order of magnitude higher than what you expect your batch sizes to be to avoid ever hitting the limit. So from that perspective, I would think that using an estimate based on the real data as an optimistic lower value would be a useful optimization.

Do we have any data on:
* the typical ratio between the typical `max.message.bytes` for the topic / avg batch size? --> to understand how big of an overestimate the `max.message.bytes` is in the typical case
* the typical number of concurrent fetch requests * `max.message.bytes` vs the available memory for fetch semaphores? --> to try to better understand how likely it is that this is going to introduce delays on the fetch path

ballard26 replied:
> perhaps we could take 2x the avg_bytes_per_offset estimate from https://github.com/redpanda-data/redpanda/pull/27909.

On thing to note is that `avg_bytes_per_offset != avg_bytes_per_batch`. I.e, consider a compacted topic where average bytes per offset would be rather low due to the sparsity of non-compacted messages in the offset space. And since the effective minimum read size from the storage layer is a batch `avg_bytes_per_offset` may not be a good proxy.

>I would expect that in general max.message.bytes will be >> the typical batch sizes seen on the topic, just because of the nature of this config. I'd expect that customers will probably typically set it at least an order of magnitude higher than what you expect your batch sizes to be to avoid ever hitting the limit. So from that perspective, I would think that using an estimate based on the real data as an optimistic lower value would be a useful optimization.

100%, I'm in favor of using real-world data to get a better estimate of batch size compared to what we could get from the config properties.

>Do we have any data on:
    - the typical ratio between the typical max.message.bytes for the topic / avg batch size? --> to understand how big of an overestimate the max.message.bytes is in the typical case
    - the typical number of concurrent fetch requests * max.message.bytes vs the available memory for fetch semaphores? --> to try to better understand how likely it is that this is going to introduce delays on the fetch path
  
Not yet, though I'm working on getting it.

ballard26 committed:
kafka/server: track average batch size in fetch metadata cache

ballard26 committed:
kafka/server: add wait time metric to fetch_memory_units_manager

ballard26 committed:
kafka/server: use avg batch size in fetch_memory_units

ballard26 committed:
kafka/server: change obligatory batch read behavior

This commit changes the obligatory batch reads to first do a non-waiting
get of memory units. Followed by a read of the partition up to the
number of acquired units in size. If that fails to read a batch then it
waits for enough units to read any batch from the partition.

dotnwat commented:
I put this in draft to remove it from some lists that were accumulating inactive PRs. Please undraft it at any time!
